Marlin/src/module/ft_motion.cpp

```diff
@@ -539,7 +539,21 @@ void FTMotion::loadBlockData(block_t * const current_block) {
 
   endPos_prevBlock += moveDist;
 
-  TERN_(FTM_HAS_LIN_ADVANCE, use_advance_lead = current_block->use_advance_lead);
+   #if FTM_HAS_LIN_ADVANCE
+    /**
+    * Use LIN_ADVANCE for blocks if all these are true:
+    * moveDist.e > 0                   : This is a print move.
+    * cfg.linearAdvK > 0               : There is an advance factor set.
+    * moveDist.e / totalLength < 3.0f  : Check for unusual high e_D ratio to detect if a retract move was
+                                        combined with the last print move due to min. steps per segment.
+                                        Never execute this with advance!
+                                        This assumes no one will use a retract length of 0mm < retr_length < ~0.2mm
+                                        and no one will print 100mm wide lines using 3mm filament or
+                                        35mm wide lines using 1.75mm filament.
+    */
+    use_advance_lead = moveDist.e > 0 && cfg.linearAdvK > 0 &&
+                       moveDist.e / (SQRT(sq(moveDist.x) + sq(moveDist.y) + sq(moveDist.z))) < 3.0f;
```

Comment:
This is so expensive. We already calculate the linear distance for a block and store it in the block, so that should be used if it is available at this point.

Meanwhile, it is better to first at least understand why a problem exists before posting a solution. The solution might not be as expensive or as redundant as this. The Simulator is a perfect tool for checking these kinds of things because we can add breakpoints in the VSCode debuggsd. The ISRs do not run as interrupts in the Simulator so we can add breakpoints anywhere we need to.

Reply:
If this is considered very expensive, then it would be a good idea to look into doing it differently in the planner.
I just copied it from there and it happens exactly as often.
My intention was to provide a quick fix for the broken linear advance on retractions and look again when I get some time (i.e now, which I did)

Reply:
We get the length of the linear move very early in the Planner process as a matter of course, so rather than re-calculate it here, we can (always?) use the stored value already in the block.